modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -295,30 +329,52 @@ def _single_batched_evaluation(
+            failure_reasons: list[tuple[str | None, dict]] = []
+            # note: interval indexes correspond to the intervals in the request
+            for interval_idx, interval_response in enumerate(response.interval_responses):
+                if interval_response.eval_aborted_reason != EvaluationAbortedReason.NOT_ABORTED:
+                    reason = get_failure_reason(interval_response.eval_aborted_reason)
+                    failure_reasons.append((reason, {}))
+                    logger.error(
+                        f"Evaluation for model {model_id_to_eval} on split {intervals[interval_idx]} "
+                        f"not started with reason: {reason}."
+                    )
+            return failure_reasons
+
+        logger.info(f"Evaluation started for model {model_id_to_eval} on intervals {intervals}.")
         self.grpc.wait_for_evaluation_completion(response.evaluation_id)
         eval_data = self.grpc.get_evaluation_results(response.evaluation_id)
         self.grpc.cleanup_evaluations([response.evaluation_id])
-        # here we assume only one interval is evaluated
-        eval_results: dict = {"dataset_size": response.interval_responses[0].dataset_size, "metrics": []}
 
-        assert len(eval_data) == 1, "The evaluation request only contains one interval so we expect only one result."
-        assert eval_data[0].interval_index == 0
-        for metric in eval_data[0].evaluation_data:
-            eval_results["metrics"].append({"name": metric.metric, "result": metric.result})
-
-        return None, eval_results
+        eval_results: list[tuple[str | None, dict[str, Any]]] = []
+        assert len(eval_data) == len(intervals), f"We expected {len(intervals)} intervals, but got {len(eval_data)}."
+        assert eval_data
+
+        # as the dataset_size comes from the `EvaluateModelResponse` and response is a dense list (no intervals are
+        # skipped), we build a list of results with the same order as the intervals.
+        # the metrics will be filled in the next loop that unwraps `EvaluationResultResponse`.
+        for interval_response in response.interval_responses:
+            if interval_response.eval_aborted_reason != EvaluationAbortedReason.NOT_ABORTED:
+                reason = get_failure_reason(interval_response.eval_aborted_reason)
+                eval_results.append((reason, {}))
+            else:
+                eval_results.append((None, {"dataset_size": interval_response.dataset_size, "metrics": []}))
+
+        for interval_result in eval_data:
+            interval_idx = interval_result.interval_index
+            assert eval_results[interval_idx][0] is None, "Evaluation failed, no metrics should be present."
+            eval_results[interval_idx][1]["metrics"] = [
+                {"name": metric.metric, "result": metric.result} for metric in interval_result.evaluation_data
+            ]
+        return eval_results
```

Comment:
Could you maybe elaborate in a comment how you construct eval_results here? I am a bit confused why there are two loops

Reply:
There are two grpc calls that one that returns the `dataset_size` and potential failure messages, the other one a sparse (only successful intervals) the metric results.

Reply:
I added a comment